[Merged by Bors] - feat: further extensions of the differential geometry elaborators

grunweg opened the pull request:
Also infer the model with corners on
- a space of continuous linear equivalences
- a closed real interval (Set.Icc)
- the complex upper half plane

And add elaborators for `HasMFDeriv(Within)At` analogous to the `mfderiv(Within)` ones.

Further cases, as well as products, disjoint unions or open subsets, will be covered in future PRs.

---

- [x] depends on: #30412

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
feat: also infer a model on a ContinuousLinearMap space

grunweg committed:
tests: infer a model with corners on a real interval

grunweg committed:
first cut: infer a model with corners on Set.Icc
Doesn't work yet, perhaps because I'm coercion too much?

grunweg committed:
tests: inferring a model on UpperHalfPlane

grunweg committed:
zeroth cut: need to import UpperHalfPlane, and get application type mismatches...

grunweg committed:
feat: elaborators for HasMFDeriv(Within)At

These are mostly analogous to those for mfderiv(Within),
with the caveat that we must coerce the second argument to a bare function.

grunweg committed:
debugging Set.Icc failures

grunweg committed:
Missing import; add right smoothness typeclasses

grunweg committed:
fix: tests for UpperHalfPlane pass now; cannot remove the import easily...

grunweg committed:
tenmp debugging

grunweg committed:
Avoid the extra import

grunweg committed:
Extract the Set.Icc from the coercion

Co-authored-by: thorimur <68410468+thorimur@users.noreply.github.com>

grunweg committed:
Fix capture of modelWithCorners

Co-authored-by: thorimur <68410468+thorimur@users.noreply.github.com>

grunweg committed:
Clean up (remove debugging output) and fix tests: we need to consider maps
between real manifolds.

grunweg committed:
Update docs

grunweg committed:
Fix

thorimur commented on Mathlib/Geometry/Manifold/Notation.lean:
```diff
@@ -329,6 +333,47 @@ where
+    --     let iTerm : Term := ← ``(𝓘($eK, $eT))
+    --     Term.elabTerm iTerm none
+    --   else
+    --     throwError "Coefficients {k} and {S} of {e} are not definitionally equal"
```
Why is this code commented out? 👀

thorimur replied:
Ah, okay! (I also finally saw the TODO above!)

I'm not sure about anything from the use case side, but seeing as it's almost identical to the branch for `ContinuousLinearMap`, it's not really adding any maintenance cost. I'd be inclined to support it! :)

grunweg replied:
So was I, and then I turned on my brain and realised that this does not make mathematical sense. (Continuous linear equivalences are not an abelian group, with addition. I've removed the TODO and added a comment why they are not handled here. They could be handled in the future, as they are a manifold, but that can happen in a future PR.)

thorimur commented on Mathlib/Geometry/Manifold/Notation.lean:
```diff
@@ -329,6 +333,47 @@ where
+  -- TODO: should this also support continuous linear equivalences?
+  fromCLM : TermElabM Expr := do
+    match_expr e with
+    | ContinuousLinearMap k S _ _ _σ _E _ _ _F _ _  _ _ =>
```
```suggestion
    | ContinuousLinearMap k S _ _ _σ _E _ _ _F _ _ _ _ =>
```

thorimur commented on Mathlib/Geometry/Manifold/Notation.lean:
```diff
@@ -329,6 +333,47 @@ where
+        -- TODO: check if σ is actually the identity!
+        let eK : Term ← Term.exprToSyntax k
+        let eT : Term ← Term.exprToSyntax e
+        let iTerm : Term := ← ``(𝓘($eK, $eT))
```
```suggestion
        let iTerm : Term ← ``(𝓘($eK, $eT))
```

thorimur commented on Mathlib/Geometry/Manifold/Notation.lean:
```diff
@@ -329,6 +333,47 @@ where
+  fromCLM : TermElabM Expr := do
+    match_expr e with
+    | ContinuousLinearMap k S _ _ _σ _E _ _ _F _ _  _ _ =>
+      if ← isDefEq k S then
```
Hmm, I want to be sure that `isDefEq` at this transparency doesn't make instance synthesis break. Can you write a test for this and for the one in `fromRealInterval` where `isDefEq` nontrivially succeeds (but `withReducible <| isDefEq` wouldn't)? E.g., wrapping in a `def`, such as using `def RealCopy := Real` for the `fromRealInterval` test.

thorimur replied:
Great! Also, if `isDefEq` turns out to be fine, we should document that we know it's fine near its invocation in a short comment. :)

grunweg replied:
I also pushed a test for the continuousLinearMap case; this already breaks earlier (at the synthesis of a NormedSpace instance for `E →L[𝕜] E''` :thinking:

thorimur reviewed (commented):
A first pass (hopefully most of the way there!); I've got to step away from my machine for a bit, but wanted to make these visible. :)

grunweg committed:
Use backticks consistently

grunweg commented on Mathlib/Geometry/Manifold/Notation.lean:
```diff
@@ -329,6 +333,47 @@ where
+        let iTerm : Term := ← ``(𝓘($eK, $eT))
+        Term.elabTerm iTerm none
+      else
+        throwError "Coefficients {k} and {S} of {e} are not definitionally equal"
```
Indeed, thanks for the reminder!

grunweg committed:
Update Mathlib/Geometry/Manifold/Notation.lean

Co-authored-by: Thomas R. Murrills <68410468+thorimur@users.noreply.github.com>

grunweg committed:
doc: note why just linear maps or continuous linear equivalences

are not handled the same way

Continuous linear equivalences are a manifold, but over a different model.

thorimur commented on MathlibTest/DifferentialGeometry/NotationAdvanced.lean:
```diff
@@ -219,17 +221,165 @@ Hint: Additional diagnostic information may be available using the `set_option d
+-- Now, with a fact about x < y: these should behave well.
+variable {x y : ℝ} [Fact (x < y)] {g : Set.Icc x y → N} {h : E'' → Set.Icc x y} {k : Set.Icc x y → ℝ}
+
+-- This test fails, but for reasons not specific to the elaborator.
```
Is this explanation correct?
```suggestion
-- This test fails, but for reasons not specific to the elaborator:
-- `Fact (0 < 2)` cannot be synthesized from `Fact (0 ≤ 2)`.
```

grunweg committed:
Apply suggestions from code review

Co-authored-by: Thomas R. Murrills <68410468+thorimur@users.noreply.github.com>

grunweg committed:
Fix test to test what it should; fix a long line

grunweg committed:
Add test for the isDefEq check

grunweg committed:
Add second test: this does not seem to test what I want!

grunweg committed:
Fix the attempted test: still fails

grunweg committed:
Fix build

grunweg committed:
Keep the blank line; complete module doc-string

grunweg commented on Mathlib/Geometry/Manifold/Notation.lean:
```diff
@@ -216,6 +216,8 @@ using the local context to infer the appropriate instance. This supports the fol
 - the model with corners on the tangent space of a manifold
 - a model with corners on a manifold, or on its underlying model space
+- a closed interval of real numbers
+- the complex upper half plane
```
```suggestion
- the complex upper half plane
- a space of continuous k-linear maps
```

grunweg committed:
Comment

grunweg committed:
Change to withReducible and update tests accordingly

grunweg commented:
I updated the defEq check and the tests: now, they are finally testing what they are supposed to test (and are failing accordingly). I've chosen to also test the trace errors (since they should indicate what is going on), and this made the diff grow by another 100 lines. But I hope this PR is now, finally, ready for good.

grunweg committed:
Merge branch 'master' into diffgeo-custom-elaborators-more

grunweg committed:
Fix
